src/tools/tools_execute.rs

```diff
@@ -232,9 +232,25 @@ async fn pp_run_tools(
                 "context_file".to_string(),
                 serde_json::to_string(&json_vec).unwrap()
             );
-            generated_other.push(message.clone());
+            let mut found_exact_same_message = false;
+            for original_msg in original_messages.iter().rev() {
+                if original_msg.role == "user" {
+                    break;
+                }
+                if original_msg.content == message.content {
+                    found_exact_same_message = true;
```

Comment:
This approach is only useful when the same context_files are in different messages. However, it does not help when model goes into a loop calling same functions again and again in one message